README.md
CTanks
======

This is a programming game, inspired by Crobots by Tom Poindexter
(http://www.nyx.net/~tpoindex/crob.html).  Players write programs for
tanks, with the intent to seek out and destroy other tanks.

Output is a JSON object, and some scripts are provided to wrap the
object up in a web page.


Included programs
-----------------

I tried to stick with the Unix philosophy of one program per task.  I
also tried to avoid doing any string processing in C.  The result is a
hodgepodge of C, Bourne shell, and awk, but at least each piece is
fairly simple to audit.


### round.sh tank1 tank2 ...

Runs a single round, awards points with rank.awk, and creates a new
summary.html with summary.awk.  This is the main interface that you want
to run from cron or whatever.
      

### forftanks tank1 tank2 ...

A program to run a round of tanks and output a JSON description of the
game.  This is what tanks.js uses to render a game graphically.
The object printed contains:

    [[game-width, game-height],
     [[tank1-color, 
      [[sensor1range, sensor1angle, sensor1width, sensor1turret],
       ...]],
      ...],
     [[
      [tank1x, tank1y, tank1angle, tank1sensangle, 
       tank1flags, tank1sensors],
      ...],
     ...]]

If file descriptor 3 is open for writes, it also outputs the results of
the round to fd3.  


### rank.awk

Processes the fd3 output of forftanks to award points and output an
HTML results table.


### summary.awk tank1 tank2

Creates summary.html, linking to all rounds and showing overall
standing.


### designer.cgi

Accepts form input and writes a tank.



Building from source
--------------------

You should be able to just run "make".  The C is supposedly ANSI C, and
might even compile on Windows.  I've built it on Linux, with glibc and
uClibc, big- and little-endian.



History
-------

This is a port of the "Tanks" program written by Paul Ferrell
<pflarr@clanspum.net> in 2009-2010.  Paul created the entire game based
off a brief description I provided him of Crobots and a vague desire to
"make something fun for high school kids to learn some programming."  We
ran Paul's Tanks as part of a 100-attendee computer security contest in
February of 2010 and by all accounts it was a huge success.  It even
made the nightly news.

Paul's version was written in Python and provided a custom language
called "Bullet", which looked like this:

    >addsensor(50, 0, 5, 1);        # 0
    >addsensor(100, 90, 150, 1);    # 1
    >addsensor(100, 270, 150, 1);   # 2
    >addsensor(100, 0, 359, 0);     # 3
    
    # Default movement if nothing is detected
                : move(70, 70) . turretccw();
    random(2, 3): move(40, 70) . turretccw();
    random(1, 3): move(70, 40) . turretccw();
    
    # We found something!!
    sense(3): move(0, 0);
    sense(1): turretcw();
    sense(2): turretccw();
    sense(0): fire();

Nick Moffitt played with this original version and convinced me (Neale)
that something like Forth would be a better language.  I added some code
to accept a scaled-down version of PostScript.  The IRC channel we
frequent collectively agreed to give this new language the derisive name
"Forf", which should ideally be followed by punching someone after it is
spoken aloud.  I wrote a Python implementation of Forf, which was slow,
and then Adam Glasgall wrote a C implementation, which was quick.

I decided to take Tanks to Def Con in July 2010, and just for bragging
rights, to have it run on an Asus WL-500gU.  This is a $50 device with a
240 MHz MIPS CPU, 16MB RAM, and a 4MB flash disk, along with an
802.11b/g radio, 4-port 10/100 switch, and an additional 10/100 "uplink"
port; it's sold as a home wireless router.  I had originally intended to
run it off a lantern battery just for fun, but eventually thought better
of it: doing so would be wasteful for no good reason.

Aaron McPhall <amcphall@mcphall.org>, my summer intern at the time, got
OpenWRT and Python onto the box and benchmarked it at about 60 seconds
for a 16-tank game, after he had profiled the code and optimized a lot
of the math.  That wasn't bad, it meant we could run a reasonably-sized
game at one turn per minute, which we knew from past experience was
about the right rate.  But it required a USB thumb drive to hold Python,
and when we used the Python Forf implementation, the run-time shot up to
several minutes.  I began this C port while Adam Glasgall, another fool
on the previously-mentioned IRC channel, started work on a C version of
a Forf interpreter.

This C version with Forf runs about 6 times faster than the Python
version with Bullet, on the WL-500gU.  A 1GHz Intel Atom runs a 16-tank
game in about 0.2 seconds.


What's so great about Forf?
---------------------------

Nothing's great about Forf.  It's a crummy language that only does
integer math.  For this application it's a good choice, for the
following reasons:

* No library dependencies, not even malloc
* Runs in fixed size memory
* Not Turing-complete, I think (impossible to make endless loops)
* Lends itself to genetic algorithms


Author
------

Neale Pickett <neale@woozle.org>


HISTORY.txt
History
=======

This is a port of the "Tanks" program written by Paul Ferrell
<pflarr@clanspum.net> in 2009-2010.  Paul created the entire game based
off a brief description I provided him of Crobots and a vague desire to
"make something fun for high school kids to learn some programming."  We
ran Paul's Tanks as part of a 100-attendee computer security contest in
February of 2010 and by all accounts it was a huge success.  It even
made the nightly news.

Paul's version was written in Python and provided a custom language
called "Bullet", which looked like this:

    >addsensor(50, 0, 5, 1);        # 0
    >addsensor(100, 90, 150, 1);    # 1
    >addsensor(100, 270, 150, 1);   # 2
    >addsensor(100, 0, 359, 0);     # 3
    
    # Default movement if nothing is detected
                : move(70, 70) . turretccw();
    random(2, 3): move(40, 70) . turretccw();
    random(1, 3): move(70, 40) . turretccw();
    
    # We found something!!
    sense(3): move(0, 0);
    sense(1): turretcw();
    sense(2): turretccw();
    sense(0): fire();

Nick Moffitt played with this original version and convinced me (Neale)
that something like Forth would be a better language.  I added some code
to accept a scaled-down version of PostScript.  The IRC channel we
frequent collectively agreed to give this new language the derisive name
"Forf", which should ideally be followed by punching someone after it is
spoken aloud.  

I decided to take Tanks to Def Con in July 2010, and just for bragging
rights, to have it run on an Asus WL-500gU.  This is a $50 device with a
240 MHz MIPS CPU, 16MB RAM, and a 4MB flash disk, along with an
802.11b/g radio, 4-port 10/100 switch, and an additional 10/100 "uplink"
port; it's sold as a home wireless router.  I had originally intended to
run it off a lantern battery just for fun, but eventually thought better
of it: doing so would be wasteful for no good reason.

Aaron McPhall <amcphall@mcphall.org>, my summer intern at the time, got
OpenWRT and Python onto the box and benchmarked it at about 60 seconds
for a 16-tank game, after he had profiled the code and optimized a lot
of the math.  That wasn't bad, it meant we could run a reasonably-sized
game at one turn per minute, which we knew from past experience was
about the right rate.  But it required a USB thumb drive to hold Python,
and when we used the Python Forf implementation, the run-time shot up to
several minutes.  I began this C port while Adam Glasgall, another fool
on the previously-mentioned IRC channel, started work on a C version of
a Forf interpreter.

This C version with Forf runs about 6 times faster than the Python
version with Bullet, on the WL-500gU.  A 1GHz Intel Atom runs a 16-tank
game in about 0.2 seconds.


What's so great about Forf?
---------------------------

Nothing's great about Forf.  It's a crummy language that only does
integer math.  For this application it's a good choice, for the
following reasons:

* No library dependencies, not even malloc
* Runs in fixed size memory
* Not Turing-complete, I think (impossible to make endless loops)
* Lends itself to genetic algorithms



brad.c
#include <stdlib.h>
#include "brad.h"

#ifndef max
#define max(a,b) (((a) > (b)) ? (a) : (b))
#define min(a,b) (((a) < (b)) ? (a) : (b))
#endif

/* Approximate vector length
 *
 * http://www.flipcode.com/archives/Fast_Approximate_Distance_Functions.shtml
 *
 * (1007/1024) * 256 == 251.75
 *  (441/1024) * 256 == 110.25
 */
bs_t
bs_approx_dist(bs_t dx, bs_t dy)
{
  bs_t x = abs(dx);
  bs_t y = abs(dy);

  return (252 * max(x, y)) + (110 * min(x, y));
}


static bs_t bs_cos_table[] = {
  256, 256, 256, 255, 255, 254, 253, 252,
  251, 250, 248, 247, 245, 243, 241, 239,
  237, 234, 231, 229, 226, 223, 220, 216,
  213, 209, 206, 202, 198, 194, 190, 185,
  181, 177, 172, 167, 162, 157, 152, 147,
  142, 137, 132, 126, 121, 115, 109, 104,
   98,  92,  86,  80,  74,  68,  62,  56,
   50,  44,  38,  31,  25,  19,  13,   6,
    0
};

bs_t
bs_cos(brad_t angle)
{
  brad_t cos;
  bs_t   a;

  a = abs(angle) % 128;
  if (a > 64) {
    a = 128 - a;
  }
  cos = bs_cos_table[a];
  a = abs(angle) % 256;
  if ((a > 64) && (a < 192)) {
    cos = -cos;
  }
  return cos;
}

bs_t
bs_sin(brad_t angle)
{
  return bs_cos(64 - angle);
}

bs_t
bs_tan(brad_t angle)
{
  return bs_sin(angle) / bs_cos(angle);
}


static brad_t bs_asin_table[] = {
   0,   0,   1,   1,   1,   2,   2,   2,
   3,   3,   3,   4,   4,   4,   4,   5,
   5,   5,   6,   6,   6,   7,   7,   7,
   8,   8,   8,   9,   9,   9,  10,  10,
  10,  11,  11,  11,  11,  12,  12,  12,
  13,  13,  13,  14,  14,  14,  15,  15,
  15,  16,  16,  16,  17,  17,  17,  18,
  18,  18,  19,  19,  19,  20,  20,  20,
  21,  21,  21,  22,  22,  22,  23,  23,
  23,  24,  24,  24,  25,  25,  25,  26,
  26,  26,  27,  27,  27,  28,  28,  28,
  29,  29,  29,  30,  30,  30,  31,  31,
  31,  32,  32,  32,  33,  33,  33,  34,
  34,  34,  35,  35,  35,  36,  36,  37,
  37,  37,  38,  38,  38,  39,  39,  39,
  40,  40,  40,  41,  41,  42,  42,  42,
  43,  43,  43,  44,  44,  45,  45,  45,
  46,  46,  46,  47,  47,  48,  48,  48,
  49,  49,  49,  50,  50,  51,  51,  51,
  52,  52,  53,  53,  53,  54,  54,  55,
  55,  55,  56,  56,  57,  57,  57,  58,
  58,  59,  59,  60,  60,  60,  61,  61,
  62,  62,  63,  63,  64,  64,  64,  65,
  65,  66,  66,  67,  67,  68,  68,  69,
  69,  70,  70,  71,  71,  72,  72,  73,
  73,  74,  74,  75,  75,  76,  76,  77,
  77,  78,  78,  79,  80,  80,  81,  81,
  82,  82,  83,  84,  84,  85,  86,  86,
  87,  87,  88,  89,  90,  90,  91,  92,
  92,  93,  94,  95,  96,  96,  97,  98,
  99, 100, 101, 102, 103, 104, 105, 106,
 108, 109, 110, 112, 114, 116, 118, 121,
 128
};

brad_t
bs_asin(bs_t n)
{
  bs_t a = abs(n);
  brad_t ret;

  if (a > 256) {
    return 0;
  }
  ret = bs_asin_table[a];
  if (n < 0) {
    ret = -ret;
  }
  return ret;
}

brad_t
bs_acos(bs_t n)
{
  return 128 - bs_asin(n);
}

brad_t
bs_atan2(bs_t y, bs_t x)
{
  bs_t   r = bs_approx_dist(x, y);
  brad_t t;

  t = bs_acos(x / r);
  if (y < 0) {
    t = 256 - t;
  }
  return t;
}


#include <math.h>
#include <stdio.h>
#define PI 3.14159265358979323846
int
main()
{
  int   i, j;
  float f;
  int   t;

  for (i = 0; i < 257; i += 1) {
    for (j = 0; j < 257; j += 1) {
      f = sqrt(i*i + j*j);
      t = bs_approx_dist(i, j);
      printf("%d  %d  %d\n", i, (int)round(f*256)/256, t/256);
    }
  }
  return 0;
}
